sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/FileAsyncApiTests.java

```diff
@@ -424,12 +433,3698 @@ public void createOptionsWithTimeToExpireRelativeToNow() {
+            .assertNext(r -> assertEquals(metadata, r.getMetadata()))
+                .verifyComplete();
+    }
+
+    @ParameterizedTest
+    @CsvSource(value = {"null,null,null,null,200", "foo,bar,fizz,buzz,200"}, nullValues = "null")
+    public void setMetadataMetadata(String key1, String value1, String key2, String value2, int statusCode) {
+        Map<String, String> metadata = new HashMap<>();
+        if (key1 != null && value1 != null) {
+            metadata.put(key1, value1);
+        }
+        if (key2 != null && value2 != null) {
+            metadata.put(key2, value2);
+        }
+
+        assertAsyncResponseStatusCode(fc.setMetadataWithResponse(metadata, null), statusCode);
+
+        StepVerifier.create(fc.getProperties())
+            .assertNext(r -> assertEquals(metadata, r.getMetadata()))
+            .verifyComplete();
+    }
+    @ParameterizedTest
+    @MethodSource("modifiedMatchAndLeaseIdSupplier")
+    public void setMetadataAC(OffsetDateTime modified, OffsetDateTime unmodified, String match, String noneMatch,
+                              String leaseID) {
+        DataLakeRequestConditions drc = new DataLakeRequestConditions()
+            .setLeaseId(setupPathLeaseCondition(fc, leaseID))
+            .setIfMatch(setupPathMatchCondition(fc, match))
+            .setIfNoneMatch(noneMatch)
+            .setIfModifiedSince(modified)
+            .setIfUnmodifiedSince(unmodified);
+
+        assertAsyncResponseStatusCode(fc.setMetadataWithResponse(null, drc), 200);
+    }
+
+    @ParameterizedTest
+    @MethodSource("invalidModifiedMatchAndLeaseIdSupplier")
+    public void setMetadataACFail(OffsetDateTime modified, OffsetDateTime unmodified, String match, String noneMatch,
+                                  String leaseID) {
+        setupPathLeaseCondition(fc, leaseID);
+        DataLakeRequestConditions drc = new DataLakeRequestConditions()
+            .setLeaseId(leaseID)
+            .setIfMatch(match)
+            .setIfNoneMatch(setupPathMatchCondition(fc, noneMatch))
+            .setIfModifiedSince(modified)
+            .setIfUnmodifiedSince(unmodified);
+
+        StepVerifier.create(fc.setMetadataWithResponse(null, drc))
+            .verifyError(DataLakeStorageException.class);
+    }
+    @Test
+    public void setMetadataError() {
+        fc = dataLakeFileSystemAsyncClient.getFileAsyncClient(generatePathName());
+
+        StepVerifier.create(fc.setMetadata(null))
+            .verifyError(DataLakeStorageException.class);
+    }
+
+    @Test
+    public void readAllNull() {
+        fc.append(DATA.getDefaultBinaryData(), 0).block();
+        fc.flush(DATA.getDefaultDataSizeLong(), true).block();
+
+        StepVerifier.create(fc.readWithResponse(null, null, null, false))
+            .assertNext(r -> {
+                r.getValue().subscribe(piece -> {
+                    TestUtils.assertArraysEqual(DATA.getDefaultBytes(), piece.array());
+                });
```

Comment:
I wouldn't recommend making a subscribe call like this in `assertNext`. Instead validate the headers in a flatMap call that creates the array.

```java
StepVerifier.create(fc.readWithResponse(null, null, null, false)
    .flatMap(r -> {
        // Verify headers here
       return FluxUtil.collectBytesInByteBufferStream(r.getValue());
    })
    .assertNext(bytes -> TestUtils.assertArraysEqual(DATA.getDefaultBytes(), bytes))
    .verifyComplete();
```